// hibiku/bts/core/db/index.js
const Promise = require('bluebird');
const mongoose = Promise.promisifyAll(require('mongoose'));
const db = mongoose.connection;


mongoose.connect('mongodb://'+settings.user.db.host+'/'+settings.user.db.name, {
    user: settings.user.db.user,
    pass: settings.user.db.pass
});
    db.on('error', console.error.bind(console, 'mongo error:'));


const pageSchema = new mongoose.Schema({
    page: String,
    fields: [{
        anchor: String,
        content: String
    }]
});

const Page = mongoose.model('Page', pageSchema);

function saveNew(data) {
    Page.findOneAndUpdate({ page: data.page }, data, {upsert: true, 'new': true}, function(err, page) {
        if (err) console.error("mongo error:", err);
    });
}

function getFields(body) {
    return Page.findOneAsync({ page: body.page });
}

module.exports = {
    save: saveNew,
    get: getFields
};
















/*const nano = Promise.promisifyAll(require('nano'));
const redis = Promise.promisifyAll(require('redis'));

https://github.com/Level/levelup
http://kyoto-client.org/
http://docs.basho.com/riak/latest/theory/why-riak/
https://github.com/ssdb/nodessdb
https://github.com/pouchdb/express-pouchdb

orientdb

http://guide.couchdb.org/editions/1/en/index.html
https://github.com/dscape/nano/blob/master/lib/nano.js
http://writings.nunojob.com/2011/08/nano-minimalistic-couchdb-client-for-nodejs.html
http://redis.io/topics/lru-cache
https://github.com/alevy/memjs
http://kkovacs.eu/cassandra-vs-mongodb-vs-couchdb-vs-redis

*/
