// hibiku/bts/core/api/posts/index.js
const Promise = require('bluebird');
const db = require('../../db');

const JSCK = require('jsck');
const fs = require('fs');
const path = require('path');

const parse = require('co-body');

const schema = JSON.parse(fs.readFileSync(path.join(__dirname,'schema.json'), 'utf-8'));
const jsck = new JSCK.draft4(schema);

function* POST() {
    if (this.request.accepts('json') && this.request.is('json')) {
        const body = yield parse.json(this.request);
        var validated = jsck.validate(body);
        if (validated.valid) {
            db.save(body);
            this.body = validated;
        } else {
            this.status = 400;
            this.body = {
                error: "bad request",
                jsck: validated
            };
        }
    } else {
        this.status = 400;
        this.body = {"error": "bad request"};
    }
}

const schemaGet = JSON.parse(fs.readFileSync(path.join(__dirname,'schemaGet.json'), 'utf-8'));
const jsckGet = new JSCK.draft4(schemaGet);

function* GET() {
    if (this.request.accepts('json') && this.request.is('json')) {
        const body = yield parse.json(this.request);
        var validated = jsckGet.validate(body);
        if (validated.valid) {
            var fields = yield db.get(body);
            fields = fields.fields.map(function (field) {
                return {
                    anchor: field.anchor,
                    content: field.content
                }
            });
            this.body = fields;
        } else {
            this.status = 400;
            this.body = {
                error: "bad request",
                jsck: validated
            };
        }
    } else {
        this.status = 400;
        this.body = {"error": "bad request"};
    }
}

module.exports = {
    post: POST,
    get: GET
};
