// hibiku/bts/core/router.js
const koa = require('koa');
const serve = require('koa-static');
const app = module.exports = koa();

const path = require('path');
const Promise = require('bluebird');
const utils = require('./utils');

const ee = require('events').EventEmitter;
const event = new ee();

const render = require('./engine');

app._listen = app.listen;
app.listen = function(port, ip) {
    port = port || 8080;
    ip = ip || '';
    event.on('listen', function() {
       console.log('\n  hibiku now listening to', [ip,port].join(':'), '\n');
       app._listen(port, ip); 
    });
};


const readFile = Promise.promisify(require('fs').readFile);
function src(path) {
    return readFile(path, 'utf-8')
}


var template = {
    'index': {
        'main title': 'Yo',
        'main content area': 'peace'
    },
    'hibiku-editor': {
        files: []
    }
};

var viewFiles = [], routes = {
    full: [],
    render: []
};
Promise.each(settings.engine.views, function(view) {
    return utils.walk(view).then(function(files) {
        files.forEach(function (file) {
            routes.full.push(file);
            if (~file.indexOf(settings.root)) {
                file = file.replace(settings.engine.extension.hibiku, '');
                routes.render.push(file.replace(settings.root, ''));
            } else {
                if (file.indexOf(settings.user.engine.extension.hibiku) === (file.length-settings.user.engine.extension.hibiku.length)) { // if it ends with the hibiku extension
                    file = file.replace(settings.user.engine.extension.hibiku, '');
                    viewFiles.push(file.replace(settings.user.root, ''));
                } else {
                    file = file.replace(settings.user.engine.extension.plain, '');
                }
                
                routes.render.push(file.replace(view, ''));
            }
        });
    });
}).then(function() {
    template['hibiku-editor'].files = viewFiles;
    
    
    var map = Object.create(null);
    var full = Object.create(null);
    routes.render.forEach(function (ro) {
        var _ro = ro;
        var hook = path.basename(ro);
        if (ro.indexOf('/bts/') === 0) {
            ro = ro.replace(/pages\/(hibiku-)?/, '');
        } else {
            var parts = ro.split(path.sep).slice(1);
            if (parts.length > 1) {
                hook = parts.join(path.sep); // hook for folders (ie. path/to/filename instead of just filename)
            }
        }
        
        map[ro] = hook;
        ~routes.full.indexOf(_ro)
        full[ro] = routes.full[routes.render.indexOf(_ro)];
    });
    
    console.log(full);
    
    map['/'] = settings.user.options.frontPage;
    
    const api = require('./api');
    api.use(app);
    
    app.use(serve(settings.root));
    app.use(serve(settings.user.root));
    
    app.use(function* () {
        var req = map[this.request.url];
        console.log(full[this.request.url]);
        if (req) {
            this.body = template[req] ? yield render(req, template[req]) :
                                        yield src(full[this.request.url]);
        } else {
            this.status = 404;
            this.body = yield render('404', { url: this.request.url });
        }
    });
    
    event.emit('listen');
});
