// hibiku/bts/core/utils.js
const Promise = require('bluebird');
const fs = require('fs');
const readdirAsync = Promise.promisify(fs.readdir);
const statAsync = Promise.promisify(fs.stat);
const path = require('path');

function walk(directory) {
    return readdirAsync(directory).map(function(file) {
        var resolved = path.join(directory, file);
        return statAsync(resolved).then(function(stat) {
            return stat.isDirectory() ? walk(resolved) : resolved;
        });
    }).reduce(function (a, b) {
        return a.concat(b);
    }, []);
}

function createRegExp(str, flags) {
  return new RegExp(str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&"), flags);
}

function eachify(obj) {
    var inspection = ({}).toString.call(obj);
    if (inspection === '[object Object]')
        throw new Error(obj + ' is not an Array');
    return obj;
}

module.exports = {
    walk: walk,
    eachify: eachify,
    regex: {
        create: createRegExp
    },
    filter: {
        truthy: function(item) { return item }
    }
}
